clang/include/clang/CIR/Dialect/IR/CIRTypes.td

```diff
@@ -94,6 +95,82 @@ def SInt16 : SInt<16>;
 def SInt32 : SInt<32>;
 def SInt64 : SInt<64>;
 
+//===----------------------------------------------------------------------===//
+// FloatType
+//===----------------------------------------------------------------------===//
+
+def CIR_FloatType : CIR_Type<"Float", "float",
+    [DeclareTypeInterfaceMethods<DataLayoutTypeInterface>]> {
+  let summary = "CIR floating-point type";
```

Comment:
Let's do this more incrementally and only focus on C/C++ regular `float` and `double` for now.

We should prevent or hide as much as possible the the ABI bits at this level and happily defer such details to LLVM lowering (or lowering prepare) unless we have concrete use cases to do so here. The rationale is that some higher level analysis might not care if something is x87 fp80 at this level, it's easier to look into `long double`s instead, if we lower this too early to the ABI representation, we might lose some flexibility.

Reply:
> We should prevent or hide as much as possible the ABI bits at this level

@bcardosolopes not sure I agree with this. At least not as an initial step.

Omitting such details would imply storing them in the IR during codegen. In the `long double` scenario, the IR might be generated with `-target x86_64-unknown-linux-gnu`, but the lowering to LLVM dialect would be unaware of this, meaning it would not know if it should generate a `f80` for X86 or `f128` for PPC.  This may be solved with a simple module-level attribute that carries this information, but involves more nuanced design decisions, making it unsuitable for this initial PR.

I believe it would be best to start emitting the types closer to what the original codegen does, then raise the abstraction level on a later PR after we have the proper infrastructure to do so.

Reply:
So do we have a consensus on this? I personally prefer keeping the binary format information in CIR as @sitio-couto said. The CIR-to-LLVM lowering code doesn't know the mapping from C/C++ floating-point types to the underlying floating-point formats and it may get trouble lowering CIR if we just keep a `float` / `double` / `long double` tag on the `cir.float` type.

Reply:
I vote that the CIR floating-point types correspond to the representation, not to the language type.

That's how it is done for integral types.  `long` can be either `s32i` or `s64i`.  `char` can be either `s8i` or `u8i`.  `wchar_t` can be `u16i`, `s32i`, or `u32i`.  CIR data structures don't record the original language type for integers, only the representation.  A `u16i` value in CIR could have been specified in the source code as `unsigned short`, `wchar_t`, or `char16_t`.

I think the users of ClangIR will be more interested in the representation and behavior of floating-point values than in the original language semantics.  For example, an optimization pass that does auto-vectorization wants to know that the values are IEEE 64-bit.  It doesn't need to know that they were originally specified as `long double` and then have to look up what format `long double` is on this platform.

Since ClangIR is an intermediate representation, there will frequently be tradeoffs where we have to choose between the high-level language representation and the low-level back end details.  In many cases we should choose the high-level language representation, because that preserves the semantics and patterns that optimizers need to identify potential optimizations.  But in the case of floating-point types, I think the low-level back end details is the much better choice.

Reply:
Thank for the feedback everyone. I think I wasn't clear enough on my last comments: the ABI should be available in CIR, but the facing type should be *named* just as we do in C/C++, `!cir.float`, `!cir.double`, etc. This means that we could have something like this for long double:

```
!long_double = !cir.ldouble<x86_fp80>
...
cir.func ... {
  ...
  %1 = cir.load %0 : cir.ptr <!long_double>, !long_double
}
```